### PR TITLE
DOC: remove seed in examples

### PR DESCRIPTION
**Pull request by tupui**
This is a follow up of #13778

Some example were not updated correctly to remove the seeding.

### CODE REVIEW

**Comment by tupui** [comment]
Already fixing the issues.

File: doc/source/tutorial/linalg.rst
Comment (tirthasheshpatel) [inline doc/source/tutorial/linalg.rst:895]: Can the `# random` directive be added at the end of the output? It looks kind of ugly here:

Screenshot

Also, does the `# may vary` comment seem more readable than just `# random`? I think the former implies the *output* may vary while it is difficult to tell *what* is random in the latter case. Is it just me?

File: doc/source/tutorial/stats/continuous_kstwo.rst
Comment (tirthasheshpatel) [inline doc/source/tutorial/stats/continuous_kstwo.rst:82]: This is even worse. Can't tell from where does the "random" comment come from. Can we shift this at the end of the output?

File: scipy/linalg/_expm_frechet.py
Comment (tirthasheshpatel) [inline scipy/linalg/_expm_frechet.py:78]:
```suggestion
```

Do we need to create the default RNG again? It was created just above this block of code. Also looks like we don't need the `scipy.linalg` import above.

File: scipy/special/add_newdocs.py
Comment (tirthasheshpatel) [inline scipy/special/add_newdocs.py:8809]: False positive? I don't think we need this here.

**Comment by tirthasheshpatel** [comment]
Are these all? I remember seeing a lot of docs using `seed` in `stats`.

**Comment by ilayn** [comment]
Wait. It's fine to remove the seed if we have a better way to fix reproducible results. I'm -1 on leaving things dangling randomly.

**Comment by tirthasheshpatel** [comment]
> Wait. It's fine to remove the seed if we have a better way to fix reproducible results. I'm -1 on leaving things dangling randomly.

I had this thought too but after #13778 got merged, I thought this is the direction we are headed to.

**Comment by ilayn** [comment]
I thought that was just the randomstate instances. I missed any seed removal on that PR then.

File: doc/source/tutorial/linalg.rst
Comment (tupui) [reply doc/source/tutorial/linalg.rst:895]: I think that it can go at the bottom yes. Also sure I can change to `# may vary`. We can actually write why. It checks that the comment starts with a pattern. There are a few examples in the doc with things like `# may vary (signs)`. I could have `# may vary (seed)` or something else.

File: doc/source/tutorial/stats/continuous_kstwo.rst
Comment (tupui) [reply doc/source/tutorial/stats/continuous_kstwo.rst:82]: Agreed, I will check.

File: doc/source/tutorial/linalg.rst
Comment (tirthasheshpatel) [reply doc/source/tutorial/linalg.rst:895]: > I think that it can go at the bottom yes. Also sure I can change to `# may vary`. We can actually write why. It checks that the comment starts with a pattern. There are a few examples in the doc with things like `# may vary (signs)`. I could have `# may vary (seed)` or something else.

That sounds good!

File: scipy/linalg/_expm_frechet.py
Comment (tupui) [reply scipy/linalg/_expm_frechet.py:78]: True for both. I did not want to change too much things around, but I can remove the import too. We are not perfectly consistent along the doc. Sometime we import NumPy, sometimes not, etc.

File: scipy/special/add_newdocs.py
Comment (tupui) [reply scipy/special/add_newdocs.py:8809]: Thanks I will check!

**Comment by tupui** [comment]
> > Wait. It's fine to remove the seed if we have a better way to fix reproducible results. I'm -1 on leaving things dangling randomly.
> 
> I had this thought too but after #13778 got merged, I thought this is the direction we are headed to. `np.random.default_rng` allows setting the seed though: `rng = np.random.default_rng(12345)`. What do you think about doing it @tupui?

From the discussion in the other PR we actually tried to not have any explicit seeding like 12345. As noted in NumPy's doc and now in the stats' tutorial of our doc, doing this initialize poorly the generator. Also, everyone uses common values like 0, 12345. Hence it leads to bias. Some people just copy-paste code (seen in production or research code all the time, regardless of the language/tech...). So how we write our examples is paramount.

**Comment by tupui** [comment]
> Are these all? I remember seeing a lot of docs using `seed` in `stats`.

I think. I did several pass of search trying to catch everything.

File: doc/source/tutorial/linalg.rst
Comment (tupui) [reply doc/source/tutorial/linalg.rst:895]: ok then I will change them all like this.

**Comment by tirthasheshpatel** [comment]
> I think. I did several pass of search trying to catch everything.

Thanks! Sorry for the `# random` to `# mamy vary` tedious change. I don't have any major comments other than that for now.

**Comment by tupui** [comment]
> > I think. I did several pass of search trying to catch everything.
> 
> Thanks! Sorry for the `# random` to `# may vary` tedious change. I don't have any major comments other than that for now.

No worries. It's easier to change than it was to add ;) I will just wait to have to go from another maintainer to make sure we are all on the same page.

**Comment by ilayn** [comment]
I get the idea in production code. But as we discussed there I don't care about the bias in the examples. If we cannot set seeds or whatever the new tech calls it, I'm against for any change since that is a big regression.

**Comment by ilayn** [comment]
In fact if there is no simple alternative to `numpy.random.seed()` that is an unfortunate change for all numpy/scipy. Besides, doctests are very important for catching regressions so this PR effectively removes test which is probably not what we want in the first place.

How we write tests are indeed paramount, but not checking what are tests are doing is not acceptable either.

File: doc/source/tutorial/arpack.rst
Comment (peterbell10) [inline doc/source/tutorial/arpack.rst:97]: Can we get reproducible results by seeding the `default_rng`?

Comment (tirthasheshpatel) [reply doc/source/tutorial/arpack.rst:97]: Yes but see https://github.com/scipy/scipy/pull/13863#issuecomment-820575709.

**Comment by tupui** [comment]
We had some discussions over some other issues (cannot find it now) where we said that docstrings should not test something which is not tested in a proper test. It should run to see that the API.

cc @rkern and @rgommers as I had most of these discussions with you.

**Comment by ilayn** [comment]
That is not related to what I wrote here. Just because some function is not the correctest way of doing the same thing any more, we cannot relax proper testing in the examples.

**Comment by tupui** [comment]
> That is not related to what I wrote here. Just because some function is not the correctest way of doing the same thing any more, we cannot relax proper testing in the examples.

I could simply check that the test coverage did not change before and after.

**Comment by ilayn** [comment]
Because they all pass before and after. The point is we won't catch them when they fail because the results are random and anything goes.

Anyways I'll let someone else look into it before things get salty which I am really not in the mood.

**Comment by tupui** [comment]
> Because they all pass before and after. The point is we won't catch them when they fail because the results are random and anything goes.
> 
> Anyways I'll let someone else look into it before things get salty which I am really not in the mood.

Sorry I upset you with this. By all means this is not what I want. I am giving a lot of time to do my best to improve SciPy as best as I can 😃

**Comment by ilayn** [comment]
No worries at all. Sometimes it is better to pull the plug on discussion the point when things get clogged and have a fresh pair of eyes instead of 1-upping. As a maintainer I can guarantee your turn will come too :)

File: doc/source/tutorial/arpack.rst
Comment (tupui) [reply doc/source/tutorial/arpack.rst:97]: And I just found back https://github.com/scipy/scipy/pull/13631#discussion_r600307550.

Comment (peterbell10) [reply doc/source/tutorial/arpack.rst:97]: I don't buy the idea that someone would copy code with random inputs. Surely they would use their own inputs if they're actually trying to achieve something other than learning the scipy function (in which case random quality is irrelevant).

However if it's really an issue, you could always use `default_rng(seed=SEED)` where `SEED` is defined in the doctest runner, so not part of the example. 
https://github.com/scipy/scipy/blob/374e5bfec63e530c34514a08d9311352138182f3/tools/refguide_check.py#L468-L469

Comment (rkern) [reply doc/source/tutorial/arpack.rst:97]: > However if it's really an issue, you could always use `default_rng(seed=SEED)` where `SEED` is defined in the doctest runner, so not part of the example.

Hmm, that would be a pleasing middle ground.

Comment (tupui) [reply doc/source/tutorial/arpack.rst:97]: Oh indeed, good point. I didn't realize this was possible. Then if it's okay I would do the change.

Comment (rgommers) [reply doc/source/tutorial/arpack.rst:97]: Removing the seeding in all examples and using `# random` to make the doctest runner happy seems fine to me. `default_rng(seed=SEED)` without defining `SEED` would make code no longer run when copy-pasting, which is too high a price to pay imho.

Comment (larsoner) [reply doc/source/tutorial/arpack.rst:97]: Ideally I think if you ran `make clean; make html` twice you'd get the same output, and with this PR you won't (better for reproducibility, smaller diffs with doc uploads to version controlled things, etc.). At the same time, I see the point about wanting to set a good example (which I agree is more important).

Another option is for for us to hack the doc building in some way to give us reproducible outputs. For example, in `conf.py` we could temporarily monkeypatch `numpy.random.default_rng` to be `lambda seed=0: original_default_rng(seed)`. That way our examples stay static, but readers just see us use `rng = numpy.random.default_rng()` in examples. I can look into this and open a PR to do this separately if people agree it solves our problems.

Comment (tupui) [reply doc/source/tutorial/arpack.rst:97]: Ark just seeing this. I started to add the SEED 😅. Well it's not for nothing if we take your approach @larsoner as I basically would have the new values for the tests... If we do this, you could commit to my branch directly. Or first we would use `# random` and then introduce this hack?

I will wait a bit for the decision then (+1 for @rgommers with the `# random`. In the end, I believe it could mean less technical debt  as we would not need to maintain changing values there.).

Just let me know 😃

Comment (larsoner) [reply doc/source/tutorial/arpack.rst:97]: > If we do this, you could commit to my branch directly. 

This seems like the easiest approach. Let's wait to see if others agree it's the best compromise.

Also I do agree with @rgommers that we will cause copy-paste problems for people with the SEED approach, so I prefer the hack-the-doc-build approach.

Comment (rgommers) [reply doc/source/tutorial/arpack.rst:97]: +1 for hack-the-doc-build, good idea

Comment (tupui) [reply doc/source/tutorial/arpack.rst:97]: ok great 😅 @larsoner I will push very soon what I did with `SEED`. Basically I added this in the `refguide_check.py`

```python
# SEED is generated with
# from numpy.random import SeedSequence
# print(SeedSequence().entropy)
DEFAULT_NAMESPACE = {'np': np, 'SEED': 1638083107694713882823079058616272161}
```
Then if I understand correctly your trick, please use this value when replacing `>>> rng = np.random.default_rng()`.

**Comment by larsoner** [comment]
Looks like there is a conflict, want to fix that?

**Comment by tupui** [comment]
> Looks like there is a conflict, want to fix that?

I think it's good now. I had some issues locally with `_ufuncs.pyx` I will clean my build and re-test. Otherwise all doctests are passing.

**Comment by tupui** [comment]
Thanks @larsoner! Let see the CI, but locally I have `OK: refguide and doctests checks passed!` 🚀

**Comment by larsoner** [comment]
From a quick look it seems like the failures are just to numerical precision in some stats test results. Might be worth making our refguide checker tolerant of these minor differences

**Comment by tupui** [comment]
@larsoner CI is green for `refguide_asv_check` 🎉

**Review by larsoner — APPROVED** [review]
LGTM +1 for merge!

**Comment by tupui** [comment]
I will send an email to the dev list once this is merged to explain how we should write new examples.

File: doc/source/tutorial/stats/plots/kde_plot3.py
Comment (peterbell10) [inline doc/source/tutorial/stats/plots/kde_plot3.py:7]: You need to pass `rng` to the `rvs` call below.

File: doc/source/tutorial/stats/plots/mgc_plot1.py
Comment (peterbell10) [inline doc/source/tutorial/stats/plots/mgc_plot1.py:21]: Needs an rng.

File: doc/source/tutorial/stats/plots/mgc_plot2.py
Comment (peterbell10) [inline doc/source/tutorial/stats/plots/mgc_plot2.py:43]: Needs an rng.

File: doc/source/tutorial/stats/plots/mgc_plot4.py
Comment (peterbell10) [inline doc/source/tutorial/stats/plots/mgc_plot4.py:44]: Both need rngs.

File: scipy/cluster/vq.py
Comment (peterbell10) [inline scipy/cluster/vq.py:419]: This example uses `np.random.multivariate_normal` below.

File: scipy/cluster/vq.py
Comment (peterbell10) [inline scipy/cluster/vq.py:690]:
```suggestion
           [ 6.17653859,  4.16719247]])
```

File: scipy/cluster/vq.py
Comment (peterbell10) [inline scipy/cluster/vq.py:696]:
```suggestion
    array([52, 27, 21])
```

File: scipy/optimize/_basinhopping.py
Comment (peterbell10) [inline scipy/optimize/_basinhopping.py:596]: If the goal is best-practice, shouldn't the rng be a class member?

File: scipy/optimize/_basinhopping.py
Comment (peterbell10) [inline scipy/optimize/_basinhopping.py:622]:
```suggestion
    ...                    niter=10, callback=print_fun, seed=rng)
```

File: scipy/optimize/_lsq/lsq_linear.py
Comment (peterbell10) [inline scipy/optimize/_lsq/lsq_linear.py:203]:
```suggestion
    >>> A = rand(m, n, density=1e-4, random_state=rng)
```

File: scipy/stats/morestats.py
Comment (peterbell10) [inline scipy/stats/morestats.py:544]: Need to fix the `rvs` calls below

File: scipy/stats/morestats.py
Comment (peterbell10) [inline scipy/stats/morestats.py:778]: Needs an rng

File: scipy/stats/morestats.py
Comment (peterbell10) [inline scipy/stats/morestats.py:854]: Fix rvs calls below.

File: scipy/stats/morestats.py
Comment (peterbell10) [inline scipy/stats/morestats.py:1506]: Fix rvs calls below.

File: scipy/stats/morestats.py
Comment (peterbell10) [inline scipy/stats/morestats.py:1593]: Fix rvs calls below.

File: scipy/stats/stats.py
Comment (peterbell10) [inline scipy/stats/stats.py:7394]: This line isn't equivalent any more since the above one used a `random_state`. Is there any way to pass it here as well?

File: tools/refguide_check.py
Comment (peterbell10) [inline tools/refguide_check.py:472]: Remove this comment.

**Review by peterbell10 — CHANGES_REQUESTED** [review]

File: scipy/optimize/_basinhopping.py
Comment (tupui) [inline scipy/optimize/_basinhopping.py:593]:
```suggestion
    >>> class MyTakeStep:
    ...    def __init__(self, stepsize=0.5):
    ...        self.stepsize = stepsize
    ...        self.rng = np.random.default_rng()
```

File: scipy/optimize/_basinhopping.py
Comment (tupui) [inline scipy/optimize/_basinhopping.py:597]:
```suggestion
    ...        x[0] += self.rng.uniform(-2.*s, 2.*s)
    ...        x[1:] += self.rng.uniform(-s, s, x[1:].shape)
```

File: tools/refguide_check.py
Comment (tupui) [inline tools/refguide_check.py:472]:
```suggestion
```
```suggestion
# SEED is generated with
# from numpy.random import SeedSequence
# print(SeedSequence().entropy)
```

**Comment by tupui** [comment]
Thank you @peterbell10 for catching these! I will try to get some time to do it during the WE. Otherwise feel free (everyone) to commit on the branch to not keep this open too long and risk further merge conflicts.

File: doc/source/tutorial/stats/plots/mgc_plot2.py
Comment (peterbell10) [inline doc/source/tutorial/stats/plots/mgc_plot2.py:47]:
```suggestion
_, _, mgc_dict = multiscale_graphcorr(x, y, random_state=rng)
```

File: doc/source/tutorial/stats/plots/mgc_plot4.py
Comment (peterbell10) [inline doc/source/tutorial/stats/plots/mgc_plot4.py:47]:
```suggestion
_, _, mgc_dict = multiscale_graphcorr(x, y, random_state=rng)
```

**Review by peterbell10 — COMMENTED** [review]
Spotted a couple more

File: scipy/stats/stats.py
Comment (tupui) [reply scipy/stats/stats.py:7394]: True, but apparently there is no way to pass a seed here. There is `args` but it's for the CDF and it's doing `*args`.

**Comment by tupui** [comment]
Thanks @peterbell10. Originally I did not look much as the plots or things which did not cause doctests to fail. I fixed these now. I also went through again `np.random.seed/rand/randint/randn/random/uniform`.

I had to put a `# may vary` for `multivariate_normal` as seeding is not sufficient as described in multiple issues (https://github.com/numpy/numpy/issues/13597, https://github.com/numpy/numpy/issues/8041).

**Review by peterbell10 — APPROVED** [review]

**Comment by tupui** [comment]
@rgommers do you want to have a final look?

**Review by rgommers — APPROVED** [review]
LGTM too and CI is all green. Let's give this a go, thanks @tupui! And thanks all reviewers!

**Comment by tupui** [comment]
Thanks Ralf 😃

**Comment by mdhaber** [comment]
Anyone else on Windows having trouble running `refguide_check` locally? I'm consistently getting a `MemoryError` in the middle of the stats refguide_check. 
Is there a verbose output option so I can tell which test is failing? `python runtests.py -g --refguide-check -v` doesn't do it.
Is there a way of running the check for a single file so I can avoid the error? `python runtests.py -g --refguide-check -t scipy.stats._entropy` doesn't do it.

**Comment by rgommers** [comment]
That doesn't sound familiar. Would be better to open a separate issue for it.

**Comment by ev-br** [comment]
> python runtests.py -g --refguide-check -t scipy.stats._entropy doesn't do it.

Sadlly, no; refguide-check only works per-submodule, so `-s stats` should work but  `-t scipy.stats._entropy` won't.

**Pull request merged**
